File: sdk-android/src/main/kotlin/dev/cloudx/sdk/EventTracker.kt
package dev.cloudx.sdk

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Handles impression, click, and viewable event reporting to the CloudX server.
 */
internal class EventTracker(
    private val client: OkHttpClient,
    private val endpoint: String
) {
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    /**
     * Report an impression event.
     * Returns true if the server accepted it, false if frequency-capped (429).
     */
    fun reportImpression(auctionId: Int, advertiserId: String, userId: String?): Boolean {
        val response = postEvent("/event/impression", auctionId, advertiserId, userId)
        return response != 429
    }

    /**
     * Report a click event.
     */
    fun reportClick(auctionId: Int, advertiserId: String, userId: String?) {
        postEvent("/event/click", auctionId, advertiserId, userId)
    }

    /**
     * Report a viewable event.
     */
    fun reportViewable(auctionId: Int, advertiserId: String, userId: String?) {
        postEvent("/event/viewable", auctionId, advertiserId, userId)
    }

    /**
     * POST an event to the given path and return the HTTP status code.
     */
    private fun postEvent(
        path: String,
        auctionId: Int,
        advertiserId: String,
        userId: String?
    ): Int {
        val json = JSONObject().apply {
            put("auction_id", auctionId)
            put("advertiser_id", advertiserId)
            if (userId != null) {
                put("user_id", userId)
            }
        }

        val requestBody = json.toString().toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url("$endpoint$path")
            .post(requestBody)
            .build()

        val response = client.newCall(request).execute()
        response.use { resp ->
            if (resp.code == 429) {
                return 429
            }
            if (!resp.isSuccessful) {
                throw CloudXException("Event $path failed: HTTP ${resp.code}")
            }
            return resp.code
        }
    }
}

File: sdk-web/src/event-tracker.ts
/**
 * Event tracking for impression, click, and viewable events.
 */
export class EventTracker {
  constructor(
    private endpoint: string,
    private publisherId?: string,
  ) {}

  /**
   * Report an impression event. Returns false if frequency-capped (429).
   */
  async reportImpression(
    auctionId: number,
    advertiserId: string,
    userId?: string,
  ): Promise<boolean> {
    const body: Record<string, unknown> = {
      auction_id: auctionId,
      advertiser_id: advertiserId,
    };
    if (userId) body.user_id = userId;
    if (this.publisherId) body.publisher_id = this.publisherId;

    const resp = await fetch(`${this.endpoint}/event/impression`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (resp.status === 429) return false;
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(text || `Impression error ${resp.status}`);
    }
    return true;
  }

  /** Report a click event. */
  async reportClick(
    auctionId: number,
    advertiserId: string,
    userId?: string,
  ): Promise<void> {
    const body: Record<string, unknown> = {
      auction_id: auctionId,
      advertiser_id: advertiserId,
    };
    if (userId) body.user_id = userId;
    if (this.publisherId) body.publisher_id = this.publisherId;

    const resp = await fetch(`${this.endpoint}/event/click`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(text || `Click error ${resp.status}`);
    }
  }

  /** Report a viewable event. */
  async reportViewable(
    auctionId: number,
    advertiserId: string,
    userId?: string,
  ): Promise<void> {
    const body: Record<string, unknown> = {
      auction_id: auctionId,
      advertiser_id: advertiserId,
    };
    if (userId) body.user_id = userId;
    if (this.publisherId) body.publisher_id = this.publisherId;

    const resp = await fetch(`${this.endpoint}/event/viewable`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(text || `Viewable error ${resp.status}`);
    }
  }
}
